Online connectivity check in connection manager

jukkar opened the pull request:
Experimenting with "online connectivity check" that I have been talking earlier. The idea is to make connection manager do an optional connectivity check (ping or http get req) to a user configurable host. If the check passes, then we have a good chance that we are connected to Internet (if the host was in the Internet) and an `NET_EVENT_CONNECTIVITY_ONLINE` mgmt event is generated. After verifying that the system is online, optional verifier can detect if we loose the connectivity, in which case `NET_EVENT_CONNECTIVITY_OFFLINE` mgmt event is generated.

Update:
This has been pending a long time but the feature should be usable atm. I created a way to test the sample application using Docker container, see `samples/net/connectivity_check` and the `README` file for details.

jukkar committed:
net: conn_mgr: Implement online connectivity status check

Add a way to check if we have an Internet connectivity before
emitting L4_CONNECTED event. We support two strategies for this,
either an ICMP Echo-Request (ping) to a user defined host, or
a HTTP GET request to a user defined URL. Both IPv4 and IPv6 checks
are supported. If the user defined host is not an IP, then DNS
is used to resolve the host address.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: mgmt: Add L4 Online Connectivity event

This online connectivity event is generated when we have
verified that we have Internet connectivity.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add online connectivity status to events

Add L4 Online Connectivity print when monitoring generated events.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: conn_mgr: Add shell command to manually trigger online check

Add "net conn check" to allow user to manually trigger online
connectivity check. This is mainly useful for debugging purposes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo_server: Start DHCP if needed

If config settings library is not enabled, which is the normal
case for non testing devices, then start DHCP by the echo-server
application if that is enabled.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo_server: Add TLS support for online checker

Add online checker callback function that should return TLS
security tags.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: conn_mgr: Add online verifier support

Add code that checks that we are still online. It uses
Trickle algorithm to trigger the re-check only if needed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: mgmt: Add L4 Offline Connectivity event

This offline connectivity event is generated when we have
verified that we cannot have Internet connectivity.
Typically this is generated when all network interfaces are
down.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo_server: Add overlay for connection manager

Add an example overlay conf file for connection manager settings.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: conn_mgr: Allow runtime configuration of used strategy

Allow user to select at runtime what online connectivity check
strategy to use.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: Add connectivity check sample application

The application can be used to verify that the online
connectivity check feature works as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
scripts: net: Add utility functions for Docker testing

Add a way to run zephyr and wait output from zephyr.exe and
decide test verdict from that.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
scripts: net: Kill zephyr.exe if user presses CTRL-C

If user killed the script using CTRL-C, the zephyr.exe was left
hanging around which caused problems when trying to run it again.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
scripts: net: Remove docker container properly

Use "docker container stop" to remove the container so that
two tests can be run in a row.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

kartben commented on samples/net/connectivity_check/README.rst:
```diff
@@ -0,0 +1,77 @@
+********
+
+This sample application enables the online connectivity checker in network
+connection manager. The checker tries to connect to user specific host either
```
```suggestion
connection manager. The checker tries to connect to user-specified host either
```

kartben commented on samples/net/connectivity_check/README.rst:
```diff
@@ -0,0 +1,77 @@
+
+This sample application enables the online connectivity checker in network
+connection manager. The checker tries to connect to user specific host either
+by a ICMP ping request, or by a HTTP(S) GET request.
```
```suggestion
by an ICMP ping request, or by a HTTP(S) GET request.
```

kartben commented on samples/net/connectivity_check/README.rst:
```diff
@@ -0,0 +1,77 @@
+by a ICMP ping request, or by a HTTP(S) GET request.
+If the connection can be established, a ``NET_EVENT_CONNECTIVITY_ONLINE`` is
+generated and application can monitor this event and then act on it.
+If the connectivity to user specified host is lost, then
```
```suggestion
If the connectivity to user-specified host is lost, then
```

kartben commented on samples/net/connectivity_check/README.rst:
```diff
@@ -0,0 +1,77 @@
+
+As the connectivity check is meant to connect to some outside network
+with public IP addresses, it might be necessary to setup network settings
+in host side so that Linux host can act as a router.
```
```suggestion
on the host side so that Linux host can act as a router.
```

kartben reviewed (commented):
quick early comments from a first pass. This will be an interesting new feature!

rlubos commented on subsys/net/lib/shell/events.c:
```diff
@@ -264,6 +264,9 @@ static const char *get_l4_desc(uint32_t event)
 	case NET_EVENT_COAP_OBSERVER_REMOVED:
 		desc = "CoAP observer removed";
 		break;
+	case NET_EVENT_CONNECTIVITY_ONLINE:
```
We shall print offline event as well.

rlubos commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -91,6 +171,799 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+
+	/* Check if we have received any data between calls to this cb. */
+	if (do_suppress) {
+		NET_DBG("No data transfer seen, not considered online.");
```
If I understand correctly, right now in case there's no communication, we'll set the state back to offline and then trigger the connectivity probing, which could report the device is online again if the probe succeeds.

Wouldn't it make sense to do this the other way around? I. e. in case there were no communication from the app activity, we trigger connectivity probing first, and then let the respective connectivity handlers (icmp/http) set the status to offline if they fail?

jukkar replied:
Yes, I suppose there are multiple ways to do this. We have the Trickle library in networking that can be used for triggering connectivity check if there is no activity seen. The criteria for what this "seen" means could be debated, it could be that the application is totally separate and only communicates in certain networks etc. in which case its activity would not count as "online". This is very much debatable and depends on use cases etc.

rlubos commented on subsys/net/conn_mgr/Kconfig:
```diff
@@ -58,4 +59,121 @@ config CONNECTIVITY_WIFI_MGMT_APPLICATION
 
 endchoice
 
+config NET_CONNECTION_MANAGER_ONLINE_CONNECTIVITY_CHECK
```
Shall we hide all options below under `if NET_CONNECTION_MANAGER_ONLINE_CONNECTIVITY_CHECK / endif` ?

jukkar replied:
There is "depends on" in those options, and they should not be available/seen if checker is not enabled.

rlubos replied:
If you want to stick to `depends on`, then it was missed for `NET_CONNECTION_MANAGER_ONLINE_CHECK_PING` and `NET_CONNECTION_MANAGER_ONLINE_CHECK_HTTP`.

rlubos reviewed (commented):
That's a lot of work, nice. Left some initial feedback.

jukkar commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -91,6 +143,537 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+	}
+
+	/* Lets assume online also for 301 "Moved Permanently" */
+	if (status == 200 || status == 301) {
```
Yes, it might be enough. I was thinking that perhaps the user of this API wants to actually connect so that the device gets registered somehow to some management system etc. This would be the "ET calls home" feature. To make that really work, the URL would need to be unique which is not atm.

jukkar replied:
> To make that really work, the URL would need to be unique which is not atm.

This was such a low hanging fruit that I implemented the support for this. Application can specify the used URL at runtime in the latest version.

glarsennordic replied:
I'm also not sure the "ET calls home" feature should really be implemented at the `conn_mgr` level anyways -- that feels squarely like application-specific behavior. IE, the application-writer (or, perhaps, cloud-library writer) should write their own "ET calls home", rather than configure `conn_mgr` to do one.

In fact, I do wonder what the intended use-case of the HTTP method in general is -- in theory, an application can get the same information by simply assuming connectivity is already there, attempting to connect, and accounting for the possibility of a connection failure/timeout, right?

So, to me, I hesitate to include such a sophisticated feature in `conn_mgr` -- it feels simpler, and more intuitive, to me, to let applications handle the rest themselves,

I am 100% on board with the ping, method, though, I think that could be useful

jukkar replied:
The HTTP support is there because you might not have PING supported in all the networks. The "ET call home" feature is just a side effect that could be utilized if needed, it was not the main motivation for this support.
The point of this PR is that each and every application does not need to implement their own way to try to connect and figure out if they are connected to Internet.

glarsennordic replied:
> The point of this PR is that each and every application does not need to implement their own way to try to connect and figure out if they are connected to Internet, or to what ever network they want to verify the connection.

I guess where I don't fully follow is that if the application needs to know this, its because the application wants to connect at some point

So why would an application use the HTTP connection check mode when they will get identical information from just connecting as usual whenever `conn_mgr` reports L4 up?

Like, why pre-check the connection, when the application could instead just simply connect?

jukkar commented on include/zephyr/net/net_event.h:
```diff
@@ -227,6 +229,12 @@ enum net_event_l4_cmd {
 #define NET_EVENT_HOSTNAME_CHANGED			\
 	(_NET_EVENT_L4_BASE | NET_EVENT_L4_CMD_HOSTNAME_CHANGED)
 
+#define NET_EVENT_CONNECTIVITY_ONLINE			\
```
Indeed, they are not documented atm. Perhaps they should be, but that needs to be done in separate PR.

kartben replied:
I don't mind documentation following up in another PR but I think making the event macros public must be addressed as part of this PR. It really doesn't sound right to introduce new sample code that uses what's right now an internal API. Thanks!

jukkar replied:
None of the events are documented atm, and you are actually the first one to notice. Anyway, I was planning to create a PR after this one is merged that documents all the events properly.

jukkar replied:
@kartben see #70443 for the event documentation

jukkar committed:
net: conn_mgr: Allow application to specify checker URL

Application can specify the used URL using the
conn_mgr_register_online_checker_url() API. If no URL is
set, then the one specified in Kconfig is used as a default.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Updated according to comments.

rlubos commented on include/zephyr/net/conn_mgr_monitor.h:
```diff
@@ -95,9 +142,98 @@ void conn_mgr_watch_l2(const struct net_l2 *l2);
+ *        that TLS support (CONFIG_NET_SOCKETS_SOCKOPT_TLS) is
+ *        enabled.
+ *
+ * @return Return 0 if URL registetration succeed, <0 otherwise.
```
```suggestion
 * @return Return 0 if URL registration succeed, <0 otherwise.
```

rlubos commented on samples/net/connectivity_check/src/main.c:
```diff
@@ -0,0 +1,146 @@
+
+	k_sem_take(&quit_lock, K_FOREVER);
+
+	printk("DONE!\n");
```
The rest of tha sample code uses log macros, perhaps we could stick to that.

jukkar replied:
The `printk()` was there on purpose so that it is only thing printed on this line. The Docker test script monitors the output and stops executing when the `DONE!` is printed.

jukkar replied:
I added a comment in the code to describe why the print is there.

rlubos commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -91,6 +172,812 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+
+#define TLS_PEER_HOSTNAME "localhost"
+		ret = zsock_setsockopt(sock, SOL_TLS, TLS_HOSTNAME,
+				       TLS_PEER_HOSTNAME,
```
This shouldn't be hardcoded, we should set the TLS hostname to the actual value the certificate was issued for (could be taken from the URL I think, or provided with the callback that feeds certs). Not setting the hostname properly will result in TLS handshake failure with real servers.

jukkar replied:
Indeed, good point. I will figure out something for the hostname.

rlubos commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -91,6 +172,812 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+	if (IS_ENABLED(CONFIG_NET_IPV4) &&
+	    online_check->hostaddr.sa_family == AF_INET) {
+		addrlen = sizeof(struct sockaddr_in);
+
```
Unneeded empty line.

rlubos reviewed (commented):
Second round of review, a bit more thorough this time. Overall looks good, it looks that the TLS hostname setting needs an update. Otherwise just some nits.

glarsennordic commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -91,6 +143,537 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+		return ret;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_SOCKOPT_TLS) && proto == IPPROTO_TLS_1_2) {
```
I would say, if we keep the HTTP connection check method, we do not want to connect. If we insist on establishing a full connection, then I have a hard time justifying this connection check method as compared to the end-user just attempting whatever connection type they ultimately intend to use with whatever credentials they ultimately intend to use

jukkar replied:
IMHO it is very important that we make a successful connection, I do not see any reason why we would not do that.
User can use non-TLS connection if certs cannot be used.

glarsennordic replied:
> I do not see any reason why we would not do that

Well, it could reduce the power used by the check, and it could also reduce the complexity of the required setup application-side. And it would also change my above argument about "why would the application want to do this?", since it might be a lower-cost pre-check than the actual connection, at least in theory.

glarsennordic commented:
I haven't properly reviewed yet, I apologize -- Will do a more detailed review tomorrow. But you have some broad thoughts from me at least for now, in comments

glarsennordic commented on subsys/net/lib/shell/conn.c:
```diff
@@ -296,6 +298,31 @@ static int cmd_net_conn(const struct shell *sh, size_t argc, char *argv[])
 }
 
+static int cmd_net_online_check(const struct shell *sh,
+				size_t argc, char *argv[])
+{
+	bool done;
+
+	done = conn_mgr_trigger_online_connectivity_check();
+	if (!done) {
+		PR("Check %sstarted.\n", "not ");
+		PR_INFO("Set %s to enable %s support.\n",
+		"CONFIG_NET_CONNECTION_MANAGER",
+		"connection manager");
+	} else {
+		PR("Check %sstarted.\n", "");
+	}
+
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(net_cmd_conn,
+	SHELL_CMD(check, NULL,
+		  "'net conn check' trigger online connectivity check.",
+		  cmd_net_online_check),
+	SHELL_SUBCMD_SET_END
+);
 
-SHELL_SUBCMD_ADD((net), conn, NULL, "Print information about network connections.",
+SHELL_SUBCMD_ADD((net), conn, &net_cmd_conn,
+		 "Print information about network connections.",
 		 cmd_net_conn, 1, 0);
```
I don't think `lib/shell/conn.c` is actually related to `conn_mgr`

I have been meaning to implement a `conn_mgr` shell that supports manual activation of several of `conn_mgr's` features. That would theoretically be a good place for this

glarsennordic replied:
Additionally, IMO, the command probably shouldn't be built if connectivity checks are not enabled

jukkar replied:
> Additionally, IMO, the command probably shouldn't be built if connectivity checks are not enabled

I have actually done net-shell differently, so that if the functionality is not there, the command is still present and it tells what to enable to make the command usable. After all the shell is mainly a debugging tool, and it can be difficult to know what commands are available if they are hidden because of some disabled option.

glarsennordic replied:
> the command is still present and it tells what to enable to make the command usable

My concern there would be a global, non-negotiable extra flash usage. As much as I do appreciate runtime usability, I'd argue the flash savings is worth making the user look up what they need in order to access a given shell.

> After all the shell is mainly a debugging tool

There are definitely some non-debug use-cases for shell to consider. For example, the TLS credentials shell, which allows run-time installation of TLS credentials.

> The purpose of this was to just to way to trigger the check for testing purposes

If that is the case, then I think it should be possible to enable SHELL and online checks without enabling this shell command. I'm happy to do that myself when I add the conn_mgr shell, assuming you are alright with that, since that shell will also be mainly for testing purposes, so all of those commands will similarly need to be optional (and disabled by default).

jukkar replied:
> There are definitely some non-debug use-cases for shell to consider. For example, the TLS credentials shell, which allows run-time installation of TLS credentials.

I slightly disagree here, imho anything requiring manual tweaking in an embedded device is basically a debug tool.

jukkar replied:
> I add the conn_mgr shell, assuming you are alright with that

I am ok with a way to control conn_mgr from the shell, but I would not like to see yet another shell module for this. So if you are doing that, please add the functionality to net-shell.

glarsennordic replied:
> but I would not like to see yet another shell module for this

Is there a serious disadvantage to it being its own shell? I tend to prefer shell modules be separated similarly to the code they govern. Since conn_mgr is a modular subsystem, IMO it stands to reason that it have its own shell module.

jukkar replied:
My argument for putting connection manager things to net-shell is that conn_mgr is directly related to various network connectivity commands already in net-shell. We could have `net cm <xxx>` for example.
There are different shells for example for wifi and ieee80215 but they are more related to controlling specific network technology instead of general network connectivity.

glarsennordic replied:
`net cm <xxx>` seems like a reasonable compromise to me. @rlubos Since you're code-owner for the networking subsystem, it seems wise to ask: do you see any potential issues with doing that?

rlubos replied:
Sounds good to me

glarsennordic reviewed (changes requested):
This is a lot of great work -- I haven't reviewed many of the later commits in much detail yet, but I have some thoughts regarding the core implementation that we should probably work out before I review those parts

Also, some of this is quite substantial, and I wonder if we could separate `conn_mgr_monitor` into multiple files.

Definitely, the two online check modes can live in their own files, so I'd request that at minimum

Possibly the trickle stuff can also be isolated into its own file? But I still haven't fully reviewed that yet.

jukkar commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -91,6 +143,537 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+	} else if (IS_ENABLED(CONFIG_NET_CONNECTION_MANAGER_ONLINE_CHECK_HTTP)) {
+		do_online_http_check(iface, URL);
+	}
+}
```
Yes, I was thinking the same. I can move all the new functionality to separate .c file for clarity.

jukkar commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -144,6 +727,8 @@ static void conn_mgr_mon_handle_update(void)
 			/* We just gained connectivity */
 			net_mgmt_event_notify(NET_EVENT_L4_CONNECTED, last_iface_up);
+
+			trigger_online_connectivity_check();
```
I am not sure if we need to fire multiple online events, I mean from the system point of view it should be enough to have online connectivity once and not do it for each interface. If there are multiple network interfaces, normally it is enough to have for example Internet connectivity via one network interface.

glarsennordic replied:
> and not do it for each interface

I definitely do not think you should fire online for each interface -- note rule 2:

> The online notification cannot fire again until all ifaces have gone offline

But I think internally Online-ness needs to be tracked for each interface

Consider the following edge case:

1. if1 and if2 become ready together, but if2 cannot reach online state
2. if1 reaches online state

Now, the ONLINE event is triggered

Now, suppose `if1` goes oper-down. In theory, our only source of online-ness has been lost.

If your trickle feature is enabled, it will probably detect this eventually, but if we tracked online-ness on a per-iface level, we could know immediately just from the oper-down state that we should now fire the OFFLINE event.

Furthermore, suppose your trickle feature is not enabled: Then if if2 never goes oper-down, then the OFFLINE event will never be fired at all.

jukkar replied:
I will shortly send a new version for review, it will not have changes in this front. I need to think more what/how to do with this.

jukkar replied:
After looking the code and the discussion above, I am not fully following the logic here. I placed the online/offline checks to the same place where `L4_CONNECTED/DISCONNECTED` events are generated. This means that the system cannot get to online state unless it has first entered `L4_CONNECTED` state.
How is the online/offline handling different from connected/disconnected handling?

glarsennordic replied:
> How is the online/offline handling different from connected/disconnected handling?

There isn't. The Connected event fires when the ***first*** iface goes oper-up. The Disconnected event fires when the ***last*** iface goes oper-down.

Similarly, the Online event should fire when the ***first*** iface goes ***online***, and the Offline event should fire when the ***last*** iface goes ***offline***.

Currently, you have it so that the Offline event can only fire when all ifaces are oper-down. So the only available ***online*** iface might go ***oper-down***, but this can end up ***not*** firing the Offline event if there is another iface that is still oper-up (but not online).

jukkar replied:
> Similarly, the Online event should fire when the _**first**_ iface goes _**online**_, and the Offline event should fire when the _**last**_ iface goes _**offline**_. (and an iface should be considered offline if it is oper-down/unready)
> 
> Currently, you have it so that the Offline event can only fire when all ifaces are oper-down. So the only available _**online**_ iface might go _**oper-down**_, but this can end up _**not**_ firing the Offline event if there is another iface that is still oper-up (but not online).

What you are describing is not how online/offline should work. As the online/offline status is tied to L4_connected/disconnected, if we get a L4_disconnected, then there cannot be any online interfaces, so offline event is triggered. Note that online/offline is just an extra property to L4 connected/disconnected, as online/offline status just does extra checks whether there is a connection is to some predefined host.

I will need to add support for checking if we were L4_connected, and get another interface L4_connected and we were offline, then we need to do online check. But if we are already online, there should be no need to do any additional online checks as one online interface should be enough.

glarsennordic replied:
> if we get a L4_disconnected, then there cannot be any online interfaces

That is true, but I am trying to point out that if you do things this way, it is possible for the only online interface to become unready (L4-down), but for L4_disconnected to not be fired

jukkar replied:
> Specifically, this happens if there is another separate iface which is ready (iface oper-up), but never passes online checks.
> 
> If that limitation is intentional, then go for it, but it seems like it is not?

Not intentional, and yes we should check what you describe. I will try to address this case next.

> IMO, if the only online iface goes L4-oper-down, it should fire L4_offline, since after that happens there are no longer any online ifaces

Yes, definitely. And I think that happens already in `conn_mgr_mon_handle_update()` which check if there are any ready interfaces and triggers L4_disconnected and offline event.

jukkar commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -91,6 +143,537 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+		       sizeof(struct online_check_data));
+
+		is_checker_init = true;
+	}
```
Hmm, true, but I faintly remember there was some reason for this. I will check what I was trying to achieve here.

jukkar commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -91,6 +143,537 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+	}
+}
+
+static void online_connectivity_check(void)
+{
+	/* We create our own copy of the iface_states in order not to keep
+	 * the lock for long time. We also check from the master states
+	 * if the interface state has changed when launching the check for
+	 * that interface.
+	 */
+	static uint16_t states[CONN_MGR_IFACE_MAX];
+	int i;
+
+	if (!IS_ENABLED(CONFIG_NET_CONNECTION_MANAGER_ONLINE_CONNECTIVITY_CHECK)) {
+		return;
+	}
+
+	k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+	memcpy(states, iface_states, sizeof(states));
+	k_mutex_unlock(&conn_mgr_mon_lock);
+
+	for (i = 0; i < ARRAY_SIZE(states); i++) {
+		if (states[i] & CONN_MGR_IF_READY) {
+			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+			if (!(states[i] & CONN_MGR_IF_READY)) {
+				k_mutex_unlock(&conn_mgr_mon_lock);
+				continue;
+			}
+
+			k_mutex_unlock(&conn_mgr_mon_lock);
+
+			/* Do connectivity check and if it succeeds,
+			 * generate CONNECTIVITY_ONLINE event. The online
+			 * check might take some time.
+			 */
+			do_online_check(conn_mgr_mon_get_if_by_index(i));
+		}
+	}
+}
```
> I don't think it is necessary to cache the states array, especially since you are locking the mutex briefly for ready ifaces anyways to check for readiness loss. At that point, I'd say just go for it and briefly check every iface, and save yourself the complexity.

I agree, it was a leftover when I had the mutex locked much longer, I will remove the cache.

jukkar commented on subsys/net/conn_mgr/Kconfig:
```diff
@@ -58,4 +59,73 @@ config CONNECTIVITY_WIFI_MGMT_APPLICATION
+	  The host can be either a IP address or a hostname.
+	  If hostname is set here, you need to enable DNS resolving by setting DNS_RESOLVER option.
+
+config NET_CONNECTION_MANAGER_ONLINE_CHECK_PING_TIMEOUT
```
Yes, it might be that having two timeouts is not really needed. We can always add one back if really needed.

jukkar commented:
Various changes according to review comments. I moved the relevant code to a separate .c file, this proved to be a nasty rebase and I hope I did not make a mistake there, in a worst case will need to send a fixup version.

kartben commented on samples/net/connectivity_check/README.rst:
```diff
@@ -0,0 +1,146 @@
+    $ scripts/net/run-sample-tests.sh samples/net/connectivity_check
+
+Note that ``run-sample-tests.sh`` script can be stopped manually by pressing
+:kbd:`CTRL+A` :kbd:`x`. The application will run couple of minutes and tries to
+verify the online connectivity.
```
```suggestion
:kbd:`CTRL+A` :kbd:`x`. The application will run for a couple of minutes and try to
verify the online connectivity.
```

jukkar commented on samples/net/connectivity_check/README.rst:
```diff
@@ -0,0 +1,146 @@
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/net/connectivity_check
+   :board: <board to use>
```
Hmm, I don't think the instructions are very qemu centric. I mention qemu once in the readme file. Perhaps we could call this emulator centric instructions as also native_sim board is mentioned.

jukkar replied:
I added more information if user wants to use some other board to run the sample.

kartben replied:
Ya, maybe emulator centric then :) My concern is with the "As the connectivity check is meant to connect to some outside network with public IP addresses, it might be necessary to setup network settings on the host side so that Linux host can act as a router." paragraph -- when I first read it I was literally like "wait, what, it said earlier I can run this sample on a supported board, so what is this referring to?"

Maybe turn the two following paragraphs into a `note` aimed at people running in an emulated environment?

jukkar replied:
I changed the chapter to `Running connectivity_check against Docker Container`, would that be enough here? Also added information that these instructions are meant for testing purposes.

kartben replied:
please push your changes tho :)

kartben reviewed (dismissed):
+1 for docs and sample part of the  PR - thanks @jukkar!

fabiobaltieri commented:
The compliance check is:

`Binary file not allowed: samples/net/connectivity_check/src/https-cert.der`

Since it should only fire if more of these are added or modified and that should not happen regularly, I think it makes sense to just bypass the check and merge without it for this PR, it should not get in the way in the future, if we does we can revisit (see https://github.com/zephyrproject-rtos/zephyr/pull/70817#issuecomment-2028148446).

jukkar committed:
net: conn_mgr: Avoid doing online check if already done

Do not do an online check for those interfaces that we have
done one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
This version has these changes:
* Go through all network interfaces that are ready and trigger online check. Make sure not to do online check multiple times for the same interface. See commit `net: conn_mgr: Avoid doing online check if already done` for details.
* Fixed merge conflicts and updated to latest main

glarsennordic commented on subsys/net/conn_mgr/online_checker.c:
```diff
@@ -801,7 +825,27 @@ void conn_mgr_online_connectivity_check(void)
+			}
+
+			check_running = true;
+			k_mutex_unlock(&check_running_lock);
```
Perhaps we should unlock `check_running_lock` after setting `states[i] |= CONN_MGR_IF_ONLINE_CHECK;`

jukkar replied:
I will leave the `check_running = true` there so that it is protected by the lock. Probably does not make much difference as you said.

glarsennordic reviewed (commented):
Sorry this took so long to re-review.

I'm not ready to approve yet, but its going in the right direction -- hopefully future reviews will be faster since I've now fully read and understood everything (I think) except your samples/test-cases

jukkar commented on include/zephyr/net/conn_mgr_monitor.h:
```diff
@@ -95,9 +148,98 @@ void conn_mgr_watch_l2(const struct net_l2 *l2);
+/** @endcond */
+
+/** Possible choices for the online connectivity check. */
+enum net_conn_mgr_online_check_type {
+	/** Use ICMP Echo-Request (ping). This is the default */
+	NET_CONN_MGR_ONLINE_CHECK_PING = 0,
+	/** Use HTTP(S) GET request. */
+	NET_CONN_MGR_ONLINE_CHECK_HTTP,
+};
```
I actually had this as Kconfig configured only at the beginning, so user was only able to either select ping or http but not both. But then realized that it actually makes it slightly cumbersome to use so made it so that user can select both if needed and made it runtime configurable. This runtime configurability can useful for example depending on application needs / networks etc.

glarsennordic replied:
But I am not convinced there is a use-case where the application actually needs to switch back and forth, and having all the settings for connectivity check be run-time configurable significantly increases the API complexity, so I think it is worth it to simplify the API and remove the unnecessary runtime customization

I just cannot think of a reason an application would want to change this from whatever was configured at build time

jukkar commented on subsys/net/conn_mgr/online_checker.c:
```diff
@@ -0,0 +1,935 @@
+
+bool conn_mgr_trigger_online_connectivity_check(void)
+{
+	if (IS_ENABLED(CONFIG_NET_CONNECTION_MANAGER_ONLINE_VERIFY)) {
+		online_check->trickle.Imin = TRICKLE_IMIN * MSEC_PER_SEC * SEC_PER_MIN;
+		online_check->trickle.Imax = TRICKLE_IMAX;
+		online_check->trickle.k = TRICKLE_K;
+	}
```
Yes, that is certainly possible.

jukkar commented on subsys/net/conn_mgr/online_checker.c:
```diff
@@ -0,0 +1,935 @@
+		online_check->trickle.k = TRICKLE_K;
+	}
+
+	if (online_check->url == NULL) {
+		conn_mgr_register_online_checker_url(ONLINE_CHECK_URL);
+	}
```
Are you thinking about possible security issues?

Having this as a build time configuration only means also that if the check target changes or is plain wrong, one would need to re-flash the device which can be a difficult task.

jukkar commented on include/zephyr/net/conn_mgr_monitor.h:
```diff
@@ -84,6 +88,55 @@ void conn_mgr_ignore_l2(const struct net_l2 *l2);
+ * @retval <0 if this online check request is cancelled and not done.
+ * @retval  0 the online check is done by the connection manager.
+ */
+typedef int (*net_conn_mgr_online_checker_t)(struct net_if *iface,
+					     const sec_tag_t **sec_tag_list,
+					     size_t *sec_tag_size,
+					     const char **tls_hostname,
+					     const char *url,
+					     const char *host,
+					     const char *port,
+					     struct sockaddr *dst,
+					     void *user_data);
```
I think it is very useful to be able to configure things at runtime, but I also see some benefits with build time configuration. I think we could let this decision to end user, so we could have a Kconfig option that allows both ways.

glarsennordic replied:
My main motivation for switching to build-time is that it would significantly reduce the complexity of certain aspects of this contribution, so if we do it both ways, we don't get that complexity reduction

jukkar replied:
Sorry but I do not see the complexity here.

jukkar commented on subsys/net/conn_mgr/conn_mgr_monitor.c:
```diff
@@ -141,10 +152,17 @@ static void conn_mgr_mon_handle_update(void)
 		if (ready_count == 0) {
 			/* We just lost connectivity */
 			net_mgmt_event_notify(NET_EVENT_L4_DISCONNECTED, last_iface_down);
+
+			conn_mgr_refresh_online_connectivity_check();
 		} else if (original_ready_count == 0) {
 			/* We just gained connectivity */
 			net_mgmt_event_notify(NET_EVENT_L4_CONNECTED, last_iface_up);
 		}
+
+		if (ready_count > 0) {
+			/* Trigger online check to be done for each ready interface */
+			conn_mgr_trigger_online_connectivity_check();
+		}
```
Currently the `conn_mgr_refresh_online_connectivity_check();` generates offline event and then re-triggers online check. There are some extra checks in `conn_mgr_online_connectivity_check()` for not triggering online check unnecessarily.
Also the `conn_mgr_trigger_online_connectivity_check()` does initialization if not done already as this might be the first time the online check functions are called.

glarsennordic replied:
My only point here is instead of having two separate places where you conditionally call `conn_mgr_trigger_online_connectivity_check`, you could just ***always*** call `conn_mgr_trigger_online_connectivity_check` at the end, since `conn_mgr_trigger_online_connectivity_check` is always called anyways


The way things are written now, it gets called if `ready_count==0`, and it also gets called if `ready_count>0`, and since ready_count cannot be negative, that means it just always gets called

glarsennordic replied:
Ah, you are right, I've been misreading `conn_mgr_refresh_online_connectivity_check` and `conn_mgr_trigger_online_connectivity_check` as the same function. I think the function names should at a minimum be updated to be a bit closer to their actual function.


I am concerned, because it seems to me, then, that the only way `NET_EVENT_CONNECTIVITY_OFFLINE` can get fired is if all ifaces are unready.


Wouldn't you want `NET_EVENT_CONNECTIVITY_OFFLINE` to fire in the edge case I pointed out?

jukkar replied:
> that the only way `NET_EVENT_CONNECTIVITY_OFFLINE` can get fired is if all ifaces are unready.

Hmm, would that be what we want here? The offline event is generated if there is no online connectivity any more.
I have envisioned this so that if there is one online interface, there is no need to trigger another online check at all. As we have only one host / URL in the configuration to check, it should not matter what interface we check.

glarsennordic replied:
> The offline event is generated if there is no online connectivity any more.

I am trying to point out that it is possible for there to be ready ifaces, but for there to be no online connectivity.

> As we have only one host / URL in the configuration to check, it should not matter what interface we check.

As long as one iface can reach that host, yes -- but it is possible for there to be at least one oper-up (ready) iface while none of the ifaces are passing online check.

glarsennordic replied:
I'll reiterate the edge-case scenario -- Imagine we are dealing with a device with two Wi-Fi interfaces, WIFIA an WIFIB.

Imagine there are two networks, NETWORKA, and NETWORKB. NETWORKA is a public network that can reach  the internet. NETWORKB is a private internal network that cannot reach the internet.

On boot, WIFIA connects to NETWORKA and becomes Oper-Up (ready). WIFIB also connects to NETWORKB and becomes oper-up (ready).

This triggers online checks. The online check for WIFIB will fail, since it cannot contact the internet. But the online check passes for WIFIA, since it can.

Now, later suppose someone accidentally unplugs the AP for NETWORKA. The WIFIA iface will detect it is no longer on that network and go oper-down.

In this scenario, the `NET_EVENT_CONNECTIVITY_OFFLINE` ***should*** be fired, since we no longer can contact the internet.

But it ***won't*** ever be fired unless WIFIB also goes down

jukkar replied:
I think we handle this case just fine with current code:

Lets assume A and B are in ready, and A is online. If now A goes down (step 1. in code snippet below), we should be able to catch that by `conn_mgr_online_connectivity_check()` which checks the states and if the interface was in READY but is no longer, then we check the online status for that interface in step 2)
```
	for (i = 0; i < state_count; i++) {
1)->		if (states[i] & CONN_MGR_IF_READY) {
			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);

			if (!(states[i] & CONN_MGR_IF_READY) ||
			    (states[i] & CONN_MGR_IF_ONLINE_CHECK)) {
				k_mutex_unlock(&conn_mgr_mon_lock);
				continue;
			}

			k_mutex_unlock(&conn_mgr_mon_lock);

			/* Do connectivity check and if it succeeds,
			 * generate CONNECTIVITY_ONLINE event. The online
			 * check might take some time.
			 */
			k_mutex_lock(&check_running_lock, K_FOREVER);

			if (check_running) {
				NET_DBG("Online check already running");
				k_mutex_unlock(&check_running_lock);
				break;
			}

			check_running = true;

			k_mutex_unlock(&check_running_lock);

			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);

			/* Tells that we are doing online check for this interface */
			states[i] |= CONN_MGR_IF_ONLINE_CHECK;

			k_mutex_unlock(&conn_mgr_mon_lock);

			/* Did we get to online? This is a synchronous call and
			 * in worst case might take a long time to return.
			 */
			ret = do_online_check(conn_mgr_mon_get_if_by_index(i));
			if (ret == 0) {
				/* We are online, no need to check the other interfaces */
				break;
			}

			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);

			states[i] &= ~CONN_MGR_IF_ONLINE_CHECK;

			k_mutex_unlock(&conn_mgr_mon_lock);
		} else {
			bool trigger_check = false;

			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);

			/* If this interface was online, then we should re-check */
2)->			if (states[i] & CONN_MGR_IF_ONLINE_CHECK) {
				trigger_check = true;
			}

			/* Clear the online check flag for non ready interface */
			states[i] &= ~CONN_MGR_IF_ONLINE_CHECK;

			k_mutex_unlock(&conn_mgr_mon_lock);

			if (trigger_check) {
				conn_mgr_refresh_online_connectivity_check();
				break;
			}
		}
	}
```
Or did I miss something?

glarsennordic replied:
Ah, yes, `conn_mgr_refresh_online_connectivity_check` does fire `NET_EVENT_CONNECTIVITY_OFFLINE`. I keep forgetting that. I think a different name for `conn_mgr_refresh_online_connectivity_check` might really make sense.

But now I am a little bit confused. Won't `conn_mgr_refresh_online_connectivity_check` get fired every time an online iface goes offline?

What if there are two online ifaces, and one goes down? In that case you would NOT want fire `NET_EVENT_CONNECTIVITY_OFFLINE`, right?

jukkar replied:
There cannot be two online interfaces, that is the whole point. If we have found out that one of the interfaces is online, no point doing another check.

glarsennordic replied:
Ahhhhh, alright. I apologize, I severely misinterpreted the intent, here. You are deliberately only ever marking a single iface as "Online" at a time.

In that case, I think this logic all works as you intend, you're right.

This makes me realize, though, that I was wrong to suggest that you use the conn_mgr states array for tracking this.

I recommend, instead, that you keep a global pointer to the iface who should be marked as `CONN_MGR_IF_ONLINE_CHECK`

This would have several advantages:

1. No longer any need to access the conn_mgr states[] array outside core code (I am pretty sure -- if not, I am sure the conn_mgr public API could be adjusted slightly to allow this)
2. Doing this would possibly let you make connectivity checks a totally separate module, too -- instead of directly firing the connectivity check functions in the conn_mgr event loop, you could instead subscribe to the L4_DISCONNECTED event to trigger `conn_mgr_trigger_online_connectivity_check`. Then you could. I'm pretty sure, just use iface oper_up and oper_down events to trigger `conn_mgr_trigger_online_connectivity_check`. But if you really want to wait until conn_mgr determines the total number of enabled ifaces has changed, we could honestly just expose that as a new event.

I do think it makes sense to keep this feature inside the conn_mgr module, but having it ***function*** as a separate component via event listeners would be a lot cleaner and a lot more modular IMO.

jukkar replied:
This PR has been pending for a long time, let me try to resurrect it.

> This makes me realize, though, that I was wrong to suggest that you use the conn_mgr states array for tracking this.

If you are referring to `states` array and the `CONN_MGR_IF_ONLINE_CHECK` bit, I think we need it in order to know which interfaces were done online check. Using a global pointer does not really cut it.

Anyway, I might have missed something as this has been too long time in stale state. I will re-submit this against latest main and lets try to get this re-reviewed and finally merged.

glarsennordic replied:
Sorry I missed your reply, here!

Alright, I will re-review from scratch when you request my review

jukkar commented:
* rebased to latest upstream

glarsennordic commented on subsys/net/conn_mgr/online_checker.c:
```diff
@@ -0,0 +1,935 @@
+
+			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+
+			states[i] &= ~CONN_MGR_IF_ONLINE_CHECK;
```
You are very careful to only ever have a single iface with the `CONN_MGR_IF_ONLINE_CHECK` flag set true at any time. (you always clear this flag here before moving on to checking the next iface).

So, why can't this just be a pointer to or index representing the iface currently marked online? Set it to -1/NULL when none are marked as online.

The only reason that I can see to use flags is if more than one iface can be marked as `CONN_MGR_IF_ONLINE_CHECK` at once, but it doesn't seem that is true?

Earlier you said:
>  I think we need it in order to know which interfaces were done online check

But if only one of these ifaces ever has `CONN_MGR_IF_ONLINE_CHECK` set at a time, I don't think there is an issue with using just a pointer or index.

I know that earlier I pushed for ***more*** flags, but at that time I severely misunderstood your goals with these additions. Now that I understand the goal, I think per-iface flags are not necessary at all

glarsennordic commented on subsys/net/conn_mgr/conn_mgr_private.h:
```diff
@@ -27,6 +27,9 @@
 #define CONN_MGR_IF_IGNORED		BIT(7)
 
+/* Online check running */
+#define CONN_MGR_IF_ONLINE_CHECK	BIT(13)
```
Though maybe this flag isn't needed at all? See my other comments

glarsennordic reviewed (commented):
Alright, I did my best to re-review from scratch, but I think we need to settle the `CONN_MGR_IF_ONLINE_CHECK` debate before we can properly discuss anything else

Overall I think the PR is pretty close. I apologize for holding you up this much

glarsennordic commented on subsys/net/conn_mgr/online_checker.c:
```diff
@@ -0,0 +1,559 @@
+static struct online_check_data online_check_data_storage;
+static struct online_check_data *online_check = &online_check_data_storage;
+
+static enum dns_resolve_status resolve_host(const char *host,
+					    const char *service,
+					    int socktype,
+					    struct zsock_addrinfo **res)
+{
+	struct zsock_addrinfo hints;
+	int ret;
+
+	if (!host || host[0] == '\0') {
+		NET_WARN("Online check hostname missing.");
+		return DNS_EAI_NONAME;
+	}
+
+	memset(&hints, 0, sizeof(hints));
+	if (IS_ENABLED(CONFIG_NET_IPV4) && !IS_ENABLED(CONFIG_NET_IPV6)) {
+		hints.ai_family = AF_INET;
+	} else if (!IS_ENABLED(CONFIG_NET_IPV4) && IS_ENABLED(CONFIG_NET_IPV6)) {
+		hints.ai_family = AF_INET6;
+	} else {
+		hints.ai_family = AF_UNSPEC;    /* Allow IPv4 or IPv6 */
+	}
+
+	hints.ai_socktype = socktype;
+
+	NET_DBG("Resolving %s", host);
+
+	ret = zsock_getaddrinfo(host, service, &hints, res);
+	if (ret != 0) {
+		NET_DBG("Cannot get %s (%d/%s)", host,
+			ret, zsock_gai_strerror(ret));
+		return ret;
+	}
+
+	return 0;
+}
+
+static int icmp_handler(struct net_icmp_ctx *ctx,
```
Can we simplify the overall architecture, and have there be a user-defined callback that just spits back an IP address and port?

Custom IP resolution strategies are unfortunately commonly necessary anyways
